button-card.js

```diff
@@ -243,33 +304,64 @@
     }
 
     _toggle(state, config) {
-      switch (config.action) {
-        case 'toggle':
-          this.hass.callService('homeassistant', 'toggle', {
-            entity_id: state.entity_id,
-          });
-          break;
-        case 'more_info': {
-          const node = this.shadowRoot;
-          const options = {};
-          const detail = { entityId: state.entity_id };
-          const event = new Event('hass-more-info', {
-            bubbles: options.bubbles === undefined ? true : options.bubbles,
-            cancelable: Boolean(options.cancelable),
-            composed: options.composed === undefined ? true : options.composed,
-          });
-          event.detail = detail;
-          node.dispatchEvent(event);
-          return event;
+      if (!config.tap_action)
+        config.tap_action = 'toggle'
+      if (config.tap_action) {
+        switch (config.tap_action.action) {
+          case 'none':
+            break;
+          case 'more-info':
+          case 'more_info': {
+            const node = this.shadowRoot;
+            const options = {};
+            const detail = { entityId: state.entity_id };
+            const event = new Event('hass-more-info', {
+              bubbles: options.bubbles === undefined ? true : options.bubbles,
+              cancelable: Boolean(options.cancelable),
+              composed: options.composed === undefined ? true : options.composed,
+            });
+            event.detail = detail;
+            node.dispatchEvent(event);
+            return event;
+          }
+          case 'location': {
+            if (!config.tap_action.navigation_path) break;
+            var root = document.querySelector("home-assistant");
+            history.pushState(null, "", config.tap_action.navigation_path);
+            const detail = {}
+            const event = new Event('location-changed', {
+              bubbles: true,
+              cancelable: false,
+              composed: true,
+            });
+            event.detail = detail;
+            root = root && root.shadowRoot;
+            root = root && root.querySelector("home-assistant-main");
+            root = root && root.shadowRoot;
+            root = root && root.querySelector("app-drawer-layout partial-panel-resolver");
+            root = root && root.shadowRoot || root;
+            root = root && root.querySelector("ha-panel-lovelace");
+            root = root && root.shadowRoot;
+            root = root && root.querySelector("hui-root");
+            root = root && root.shadowRoot;
+            root = root && root.querySelector("ha-app-layout #view");
+            root = root && root.firstElementChild;
+            if (root) root.dispatchEvent(event);
```

Comment:
Why the extract checks here for shadowRoot as opposed to `more-info`?

Might want to take a look at my code for the [radial-menu](https://github.com/custom-cards/radial-menu/blob/master/dist/radial-menu.js#L2546)

Hopefully we can get HA core to expose these functions at some point in the future so each card isn't rewriting this stuff :)

The main take away, is that node is just `this`

Reply:
I'm kind of learning so I checked some other code on how it was done and copied it here 😊 
But it's now fixed